temporalio/workflow.py

```diff
@@ -143,10 +143,38 @@ def decorator(cls: ClassType) -> ClassType:
     return decorator
 
 
+def init(
+    init_fn: CallableType,
+) -> CallableType:
+    """Decorator for the workflow init method.
+
+    This may be used on the __init__ method of the workflow class to specify
+    that it accepts the same workflow input arguments as the ``@workflow.run``
+    method. It may not be used on any other method.
+
+    If used, the workflow will be instantiated as
+    ``MyWorkflow(**workflow_input_args)``. If not used, the workflow will be
+    instantiated as ``MyWorkflow()``.
+
+    Note that the ``@workflow.run`` method is always called as
+    ``my_workflow.my_run_method(**workflow_input_args)``. Therefore, if you use
+    the ``@workflow.init`` decorator, the parameter list of your  __init__ and
```

Comment:
> the parameter list of your  __init__ and `@workflow.run` methods will usually be identical.

I think we should require/enforce the params are identical if this decorator is present and the run method is present

Reply:
OK. Yes, I was tentatively thinking of not having a contract without any static (decorator-time) constraints:

> If you use the decorator, then it is your responsibility to ensure that your __init__ signature is such that it will accept the arguments passed at workflow-run time. As long as that is true, the parameter signatures of the two methods can be whatever Python allows.

But I can't think of a clear counter-example where the stricter contract would make desirable code invalid, whereas enforcing identical params at decorator-eval time has the benefit of catching incorrect code at import time, and also that it can be relaxed in the future if we want to (whereas the looser contract cannot easily be made stricter).

I do think we want to allow names to differ, e.g. for this scenario where an underscore (perhaps required by a linter or styleguide) is used to mark an unused parameter.

```python
@workflow.defn
class MyWorkflow:
    @workflow.init
    def __init__(self, arg: str) -> None:
        self.value = arg

    @workflow.run
    async def run(self, _arg: str) -> str:
        return self.value
```

1bebf37927bc067de7df3bf2f04958c318a78e3f